icu4c/source/i18n/rbnf.cpp

```diff
@@ -1345,7 +1230,7 @@ RuleBasedNumberFormat::parse(const UnicodeString& text,
                              Formattable& result,
                              ParsePosition& parsePosition) const
 {
-    if (!fRuleSets) {
+    if (!fRuleSets || unparseable) {
```

Comment:
If the entire set of rules are unparseable, fail immediately. While I could technically traverse all possible paths where a bad rule could be used and only mark those as unparseable, I took the easy route, and failed earlier.